File: app/src/main/java/com/example/ruslanyussupov/popularmovies/data/DataSource.java
package com.example.ruslanyussupov.popularmovies.data;



import com.example.ruslanyussupov.popularmovies.data.model.Movie;
import com.example.ruslanyussupov.popularmovies.data.model.Review;
import com.example.ruslanyussupov.popularmovies.data.model.Video;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.Single;

public interface DataSource {

    Observable<List<Movie>> getMovies(Filter filter, int page);

    Observable<List<Video>> getMovieTrailers(int movieId);

    Observable<List<Review>> getMovieReviews(int movieId);

    void deleteFromFavourite(Movie movie);

    void addToFavourite(Movie movie);

    Single<Movie> getFavouriteMovie(int movieId);

    enum Filter {
        POPULAR, TOP_RATED, FAVOURITE
    }

}

File: app/src/main/java/com/example/ruslanyussupov/popularmovies/data/DataSource.kt
package com.example.ruslanyussupov.popularmovies.data


import androidx.paging.PagedList
import com.example.ruslanyussupov.popularmovies.data.model.*

interface DataSource {

    suspend fun getMovies(filter: Filter, pageSize: Int = 10): Listing<PagedList<Movie>>

    suspend fun getMovieVideos(filter: Filter, movieId: Int): Listing<List<Video>>

    suspend fun getMovieReviews(filter: Filter, movieId: Int, pageSize: Int = 10): Listing<PagedList<Review>>

    suspend fun deleteFavourite(movie: Movie)

    suspend fun insertFavourite(movie: Movie)

    suspend fun getFavourite(movieId: Int): Favorite?

    suspend fun savePopularMovies(movies: List<Movie>)

    suspend fun saveTopRatedMovies(movies: List<Movie>)

    suspend fun saveReviews(movieId: Int, reviews: List<Review>)

    suspend fun saveVideos(movieId: Int, videos: List<Video>)

    suspend fun deletePopularMovies()

    suspend fun deleteTopRatedMovies()

    suspend fun deleteReviews(movieId: Int)

    suspend fun deleteVideos(movieId: Int)

    enum class Filter {
        POPULAR, TOP_RATED, FAVOURITE
    }

}
